File: services/backend/app/schemas/privacy.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


class PersonalDataExportResponse(BaseModel):
    generated_at: datetime
    user: dict[str, object]
    training_policy: dict[str, object] | None
    sessions: list[dict[str, object]]
    defect_profiles: list[dict[str, object]]
    weekly_reports: list[dict[str, object]]
    appeals: list[dict[str, object]]
    duplicate_complaints: list[dict[str, object]]
    privacy_audit_events: list[dict[str, object]]


class TrainingDeletionResponse(BaseModel):
    session_id: UUID
    deleted: bool
    counts: dict[str, int]


class AccountDeletionResponse(BaseModel):
    request_id: UUID
    proof_code: str
    status: Literal["QUEUED"]


class DeletionStatusRequest(BaseModel):
    request_id: UUID
    proof_code: str = Field(min_length=8, max_length=128)

    @field_validator("proof_code")
    @classmethod
    def strip_proof_code(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("proof_code must not be blank")
        return stripped


class DeletionStatusResponse(BaseModel):
    request_id: UUID
    status: Literal["QUEUED", "RUNNING", "SUCCEEDED", "FAILED"]
    counts: dict[str, object]
    error_code: str | None
    requested_at: datetime
    started_at: datetime | None
    completed_at: datetime | None

File: apps/web/src/api/privacy.ts
import { authHeaders, requestJson } from "./client";

export interface WeeklyReportResponse {
  id: string;
  week_start: string;
  week_end: string;
  metrics: Record<string, unknown> & {
    completed_session_count?: number;
  };
  summary: string;
  created_at: string;
  updated_at: string;
}

export interface PersonalDataExportResponse {
  generated_at: string;
  user: Record<string, unknown>;
  training_policy: Record<string, unknown> | null;
  sessions: Record<string, unknown>[];
  defect_profiles: Record<string, unknown>[];
  weekly_reports: Record<string, unknown>[];
  appeals: Record<string, unknown>[];
  duplicate_complaints: Record<string, unknown>[];
  privacy_audit_events: Record<string, unknown>[];
}

export interface TrainingDeletionResponse {
  session_id: string;
  deleted: boolean;
  counts: Record<string, number>;
}

export interface AccountDeletionResponse {
  request_id: string;
  proof_code: string;
  status: "QUEUED";
}

export interface DeletionStatusResponse {
  request_id: string;
  status: "QUEUED" | "RUNNING" | "SUCCEEDED" | "FAILED";
  counts: Record<string, unknown>;
  error_code: string | null;
  requested_at: string;
  started_at: string | null;
  completed_at: string | null;
}

export async function fetchWeeklyReports(accessToken: string): Promise<WeeklyReportResponse[]> {
  return requestJson<WeeklyReportResponse[]>("/api/v1/me/weekly-reports", {
    headers: authHeaders(accessToken),
  });
}

export async function exportPersonalData(
  accessToken: string,
): Promise<PersonalDataExportResponse> {
  return requestJson<PersonalDataExportResponse>("/api/v1/me/export", {
    headers: authHeaders(accessToken),
  });
}

export async function deleteTraining(
  accessToken: string,
  sessionId: string,
): Promise<TrainingDeletionResponse> {
  return requestJson<TrainingDeletionResponse>(`/api/v1/me/trainings/${sessionId}`, {
    method: "DELETE",
    headers: authHeaders(accessToken),
  });
}

export async function deleteAccount(accessToken: string): Promise<AccountDeletionResponse> {
  return requestJson<AccountDeletionResponse>("/api/v1/me", {
    method: "DELETE",
    headers: authHeaders(accessToken),
  });
}

export async function fetchDeletionStatus(
  requestId: string,
  proofCode: string,
): Promise<DeletionStatusResponse> {
  return requestJson<DeletionStatusResponse>("/api/v1/privacy/deletion-status", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ request_id: requestId, proof_code: proofCode }),
  });
}
